actnow/profiles/views.py

```diff
@@ -0,0 +1,12 @@
+from rest_framework.permissions import IsAuthenticated
+from rest_framework.viewsets import ModelViewSet
+
+from .models import UserProfile
+from .permissions import IsOwnerOrReadOnly
+from .serializers import UperProfileSerializer
+
+
+class UserProfileView(ModelViewSet):
+    queryset = UserProfile.objects.all()
+    serializer_class = UperProfileSerializer
+    permission_classes = [IsAuthenticated, IsOwnerOrReadOnly]
```

Comment:
**`IsAuthenticated`** checks for `is_authenticated` and yet **`IsOwnerOrReadOnly`** again checks for `is_authenticated`. What do we really want to check for?

Reply:
The `is_authenticated` check in `IsOwnerOrReadOnly` is for the `profile` object i.e actions like `edit, delete, get(single object)` (it is implemented in the `has_object_permission` method).
The `IsAuthenticated` is for the profiles endpoint i.e for actions like `list` (it is implemented in the `has_permission` method)
What we want to check is an authenticated user can `view/list` other users profile. This is achieved by the check inside 
```
if request.method in SAFE_METHODS:
    return bool(request.user.is_active and request.user.is_authenticated)
```

Reply:
To make sure I get you, what will happen if we only have `IsOwnerOrReadOnly` in the `permission_classes`?

Reply:
If we only have `IsOwnerOrReadOnly`, any user(including no authenticated) will be able to 
1. View all users directly from `/profiles/users/`
2. Create User profiles from the same endpoint.

The only thing they won't be able to do is to perform edits/delete of an individual user profile due to the `has_object_permission` check

Reply:
So how does it know that for **list** use `IsAuthenticated`, for others (**create**, **retrieve**, **update**, etc.) use `IsOwnerOrReadOnly`?

Reply:
@esirK I also think `isOwnerOrReadOnly` is sufficient here (covers all cases)

Reply:
@kilemensi  @KhadijaMahanga 
The `has_permissions` are run against all incoming requests. i.e get, post e.t.c. But to get `object-level-permissions` in our case(ensuring only a profile owner can edit/delete a user profile), we need to implement the `has_object_permission` You can check out [these](https://www.django-rest-framework.org/api-guide/permissions/#examples) examples for some more clarification. So the `isOwnerOrReadOnly` will use the default `has_permissions` method from the `BasePermission` class and return `True` by default for all requests that aren't object specific.

Reply:
To get the `isOwnerOrReadOnly` working alone, we could add a  `has_permissions` method to it that does the `request.user.is_authenticated` check

Reply:
@esirK yeah, that's what I wanted to suggest. We could add

```
def has_permission(self, request, view):
        return request.method in SAFE_METHOD
```

Reply:
The only issue with this @KhadijaMahanga  is it will allow `POSTs` from unauthenticated users. We also only want to let only authenticated users view other users. So
maybe 
```
def has_permission(self, request, view):
        return request.user.is_authenticated
```
??

Reply:
I think lets decide how are going to be handling permissions as well as how to name and write the permission classes.

1. Are we going to tie this into Django model permissions (and may be django-guardian)?
2. Are we combining or separating View level and Object level permissions classes?
3. How are we [specifying](https://www.django-rest-framework.org/api-guide/permissions/#overview-of-access-restriction-methods) these permissions on View classes?

Reply:
@esirK  correct!

Reply:
What's the verdict here @esirK  / @KhadijaMahanga ?

Reply:
Ola @kilemensi 
For 1, I'm not certain I totally get what you mean 
2. I think it's better to keep View level permission separate from Object level permissions. Like how we are currently doing it.
3.  We are currently specifying using `permission_classes` approach.

Reply:
1. Django allows you to [customise model permissions](https://docs.djangoproject.com/en/3.2/topics/auth/customizing/#custom-permissions) and DRF can tap into [that](https://www.django-rest-framework.org/api-guide/permissions/#djangomodelpermissions)

Reply:
@kilemensi 
I see. Right now we aren't tieng permissions to a model. IMO, I think using object permissions might be overkill. I feel like determining permissions is a logic that should be placed in the views. 
What do you think @KhadijaMahanga